rhino/src/main/java/org/mozilla/javascript/ScriptRuntime.java

```diff
@@ -194,19 +197,15 @@ public static ScriptableObject initSafeStandardObjects(
         NativeJavaObject.init(scope, sealed);
         NativeJavaMap.init(scope, sealed);
 
-        boolean withXml =
-                cx.hasFeature(Context.FEATURE_E4X) && cx.getE4xImplementationFactory() != null;
-
         // define lazy-loaded properties using their class name
         new LazilyLoadedCtor(
                 scope, "Continuation", "org.mozilla.javascript.NativeContinuation", sealed, true);
 
-        if (withXml) {
-            String xmlImpl = cx.getE4xImplementationFactory().getImplementationClassName();
-            new LazilyLoadedCtor(scope, "XML", xmlImpl, sealed, true);
-            new LazilyLoadedCtor(scope, "XMLList", xmlImpl, sealed, true);
-            new LazilyLoadedCtor(scope, "Namespace", xmlImpl, sealed, true);
-            new LazilyLoadedCtor(scope, "QName", xmlImpl, sealed, true);
+        if (cx.hasFeature(Context.FEATURE_E4X)) {
+            XMLLoader loader = loadOneServiceImplementation(XMLLoader.class);
+            if (loader != null) {
+                loader.load(scope, sealed);
+            }
```

Comment:
CHECKME: Should we log some warning in future, if `FEATURE_E4X` is enabled (unfortunately, it is by default), but no XMLLoader found?

Reply:
That's an interesting point. Unfortunately we don't have a way to log a warning and I'm reluctant to add a dependency on a logging package or just "System.out.println", and throwing an exception will probably break someone, so I'm not sure realistically what we can do -- any ideas?

Reply:
In my other PR I had the idea to deprecate the flag. One idea is to enable xml parsing only, when the implementation is on class path.

But for now, I'm happy if this will get merged, as it does not introduce a functional change. I agree, that printing a warning could break things for others. We have the same situation in master branch, that we can parse, but not use xml, if dom3 is not present. And there is also no warning